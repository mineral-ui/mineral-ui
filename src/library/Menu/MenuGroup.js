/* @flow */
import React from 'react';
import { createStyledComponent } from '../styles';
import MenuGroupTitle from './MenuGroupTitle';

type Props = {
  /** [MenuItems](/components/menu-item) */
  children?: React$Node,
  /** Title for the group */
  title?: React$Node
};

export const componentTheme = (baseTheme: Object) => ({
  MenuGroup_margin: baseTheme.space_stack_sm,

  ...baseTheme
});

const Root = createStyledComponent(
  'div',
  props => {
    const theme = componentTheme(props.theme);

    return {
      margin: `${theme.MenuGroup_margin} 0`,
      '&:first-child,& + &': {
        marginTop: 0
      },

      '&:last-child': {
        marginBottom: 0
      }
    };
  },
  {
    displayName: 'MenuGroup'
  }
);

/**
 * MenuGroups assemble [MenuItems](/components/menu-item) and can display a title.
 * Grouping Menu options provides context clues to users about related actions.
 * An optional title can be added to reinforce the intent of the grouping.
 */
export default function MenuGroup({ children, title, ...restProps }: Props) {
  return (
    <Root {...restProps}>
      {title && <MenuGroupTitle>{title}</MenuGroupTitle>}
      {children}
    </Root>
  );
}
