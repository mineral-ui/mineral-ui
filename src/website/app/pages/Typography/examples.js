/* @flow */
import React from 'react';
import { createStyledComponent } from '../../../../library/styles';
import Text from '../../../../library/Text';

type TypeExample = {
  description: string,
  examples: Array<Object>
};

type TypeExamples = Array<TypeExample>;

const styles = {
  list: ({ theme }) => ({
    fontSize: theme.fontSize_ui,
    listStyle: 'square',
    paddingLeft: theme.space_inline_lg
  })
};

const List = createStyledComponent('ul', styles.list);
const examples: TypeExamples = [
  {
    description: `### Long-form Text

Default text and caption styles are intended for sections of content that
provide description and assistance. **Paragraph text** is primarily used for
content sections. **Caption Text** is used to highlight or offset content, such
as inline help or when captioning dashboard content.`,
    examples: [
      {
        content: <Text>Paragraph</Text>,
        value: '<p>, <li>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text appearance="prose">Paragraph (Prose)</Text>,
        value: '<p.prose>',
        color: 'color_gray_100',
        font: 'Regular (16/1)'
      },
      {
        content: <Text appearance="mouse">Paragraph (Mouse)</Text>,
        value: '<p.mouse>',
        color: 'color_gray_100',
        font: 'Regular (11/0.6875)'
      },
      {
        content: (
          <List>
            <li>Alpha</li>
            <li>Bravo</li>
            <li>Charlie</li>
          </List>
        ),
        // the design notes have <ol> here as well,
        // but list-style will override any numbers shown on <ol> list items
        value: '<ul>, <ol>',
        color: 'color_gray_100',
        font: 'Regular (14/0.875)'
      },
      {
        content: <Text element="h5">Caption - Strong</Text>,
        value: '<h5>',
        color: 'color_gray_100',
        font: 'Bold (14/0.875)'
      },
      {
        content: <Text element="h6">Caption - Light</Text>,
        value: '<h6>',
        color: 'color_gray_80',
        font: 'Regular (14/0.875)'
      }
    ]
  },
  {
    description: `### Display Text

Use **Display text** sparingly to make a bold statement, or draw attention to
different sections in an application view. Display text is for short titles and
section divisions—it’s not for long-form content.`,
    examples: [
      {
        content: <Text element="h1">Page Title</Text>,
        value: '<h1>',
        color: 'color_gray_100',
        font: 'ExtraBold (34/2.125)'
      },
      {
        content: <Text element="h2">Large Section Header</Text>,
        value: '<h2>',
        color: 'color_gray_80',
        font: 'Bold (28/1.75)'
      },
      {
        content: <Text element="h3">Medium Section Header</Text>,
        value: '<h3>',
        color: 'color_gray_80',
        font: 'Bold (22/1.375)'
      },
      {
        content: <Text element="h4">Small Section Header</Text>,
        value: '<h4>',
        color: 'color_gray_80',
        font: 'Bold (18/1.125)'
      }
    ]
  }
];

export default examples;
