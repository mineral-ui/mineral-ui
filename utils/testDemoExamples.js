/* @flow */
import React from 'react';
import { mount } from 'enzyme';
import { LiveProvider, LivePreview } from 'react-live';
import ThemeProvider from '../src/library/themes/ThemeProvider';

type Example = {
  title: React$Node,
  scope?: Object,
  id: string,
  source?: string
};
type Examples = Array<Example>;
type Options = {
  exclude?: Array<string> // Example id
};

export default function testDemoExamples(
  examples: Examples,
  options: Options = {}
) {
  if (options.exclude) {
    const exclusions = options.exclude || [];
    examples = examples.filter(example => !exclusions.includes(example.id));
  }

  return describe('demo examples', () => {
    examples.filter(({ scope, source }) => scope && source).map(example => {
      it(example.id, () => {
        const component = mount(
          <ThemeProvider>
            <LiveProvider
              code={example.source}
              scope={example.scope}
              mountStylesheet={false}>
              <LivePreview />
            </LiveProvider>
          </ThemeProvider>
        );
        expect(component).toMatchSnapshot();
      });
    });
  });
}
