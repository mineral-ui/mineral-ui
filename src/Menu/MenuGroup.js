/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { createStyledComponent } from '../utils';
import MenuGroupTitle from './MenuGroupTitle';

type Props = {
  /** [MenuItems](./menu-item) */
  children?: React$Node,
  /** Title for the group */
  title?: React$Node
};

const Root = createStyledComponent(
  'div',
  ({ theme }) => ({
    margin: `${theme.spacing_single} 0`,
    '&:first-child,& + &': {
      marginTop: 0
    },

    '&:last-child': {
      marginBottom: 0
    }
  }),
  {
    displayName: 'MenuGroup'
  }
);

/**
 * MenuGroups group together [Menu Items](./menu-items) and can display a title.
 */
export default function MenuGroup({ children, title, ...restProps }: Props) {
  return (
    <Root {...restProps}>
      {title && <MenuGroupTitle>{title}</MenuGroupTitle>}
      {children}
    </Root>
  );
}
