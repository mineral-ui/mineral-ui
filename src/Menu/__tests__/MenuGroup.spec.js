/* @flow */
import React from 'react';
import { shallow } from 'enzyme';
import MenuGroup from '../MenuGroup';
import examples from '../../website/app/demos/Menu/examples/MenuGroup';
import testDemoExamples from '../../../utils/testDemoExamples';

function shallowMenuGroup(props = {}) {
  const menuGroupProps = {
    ...props
  };

  return shallow(<MenuGroup {...menuGroupProps} />);
}

describe('MenuGroup', () => {
  describe('renders', () => {
    it('root', () => {
      const menuGroup = shallowMenuGroup();

      expect(menuGroup.exists()).toEqual(true);
    });
  });

  testDemoExamples(examples);
});
