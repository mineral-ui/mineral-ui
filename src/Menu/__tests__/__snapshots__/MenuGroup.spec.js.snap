// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuGroup demo examples menu-group 1`] = `
.glamor-25,
[data-glamor-25] {
  overflow: hidden;
}

.glamor-25 > div,
[data-glamor-25] > div {
  background-color: #fff;
  float: left;
  margin: 0 0.5rem 0.5rem 0;
  width: 14em;
}

.glamor-24,
[data-glamor-24] {
  box-sizing: border-box;
  color: #333840;
  font-family: Open Sans, -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.25;
  outline: 0;
}

.glamor-24 *,
[data-glamor-24] *,
.glamor-24 *::before,
[data-glamor-24] *::before,
.glamor-24 *::after,
[data-glamor-24] *::after {
  box-sizing: inherit;
}

.glamor-11,
[data-glamor-11] {
  margin: 0.5em 0;
}

.glamor-11:first-child,
[data-glamor-11]:first-child,
.glamor-11 + .glamor-11,
[data-glamor-11] + [data-glamor-11] {
  margin-top: 0;
}

.glamor-11:last-child,
[data-glamor-11]:last-child,
.glamor-11[data-simulate-lastchild],
[data-glamor-11][data-simulate-lastchild] {
  margin-bottom: 0;
}

.glamor-0,
[data-glamor-0] {
  font-size: 0.75em;
  font-weight: 700;
  margin: 0;
  padding: 1.3333333333333333em 1.3333333333333333em 0.6666666666666666em;
}

.glamor-5,
[data-glamor-5] {
  color: #333840;
  cursor: pointer;
  font-weight: 400;
  padding: 0.5em 1em;
}

.glamor-5:focus,
[data-glamor-5]:focus,
.glamor-5[data-simulate-focus],
[data-glamor-5][data-simulate-focus] {
  background-color: #ebeff5;
  outline: 0;
}

.glamor-5:hover,
[data-glamor-5]:hover,
.glamor-5[data-simulate-hover],
[data-glamor-5][data-simulate-hover] {
  background-color: #ebeff5;
}

.glamor-5:active,
[data-glamor-5]:active,
.glamor-5[data-simulate-active],
[data-glamor-5][data-simulate-active] {
  background-color: #c8d1e0;
}

.glamor-5 [role="img"],
[data-glamor-5] [role="img"] {
  box-sizing: content-box;
  display: block;
  fill: #2e6fd9;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.glamor-5 [role="img"]:first-child,
[data-glamor-5] [role="img"]:first-child {
  margin-right: 0.5em;
}

.glamor-5 [role="img"]:last-child,
[data-glamor-5] [role="img"]:last-child {
  margin-left: 0.5em;
}

.glamor-4,
[data-glamor-4] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.glamor-3,
[data-glamor-3] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex: 1 1 auto;
  flex-wrap: wrap;
  font-size: 0.875em;
  justify-content: space-between;
  padding: 0.21428571428571427em 0 0.2857142857142857em;
  white-space: normal;
  word-break: break-all;
  -webkit-flex: 1 1 auto;
  -webkit-box-lines: multiple;
  -webkit-flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.glamor-1,
[data-glamor-1] {
  margin-right: 0.5714285714285714em;
  word-break: break-word;
}

.glamor-2,
[data-glamor-2] {
  color: #58606e;
  font-size: 0.75em;
  padding-top: 0.25em;
  word-break: break-word;
}

<ThemeProvider>
  <ThemeProvider>
    <LiveProvider>
      <div
        className="react-live"
        scope={
          Object {
            "DemoLayout": [Function],
            "Menu": [Function],
            "MenuGroup": [Function],
            "MenuItem": [Function],
          }
        }
      >
        <LivePreview>
          <div
            className="react-live-preview"
          >
            <glamorous(div)>
              <div
                className="glamor-25"
              >
                <Menu>
                  <Menu>
                    <div
                      className="glamor-24"
                    >
                      <MenuGroup
                        title="Menu Group Title"
                      >
                        <MenuGroup>
                          <div
                            className="glamor-11"
                          >
                            <MenuGroupTitle>
                              <MenuGroupTitle>
                                <h3
                                  className="glamor-0"
                                >
                                  Menu Group Title
                                </h3>
                              </MenuGroupTitle>
                            </MenuGroupTitle>
                            <MenuItem>
                              <MenuItem
                                disabled={undefined}
                                onClick={undefined}
                                onKeyDown={undefined}
                                tabIndex={0}
                                variant="regular"
                              >
                                <div
                                  className="glamor-5"
                                  onClick={undefined}
                                  onKeyDown={undefined}
                                  tabIndex={0}
                                >
                                  <glamorous(span)>
                                    <span
                                      className="glamor-4"
                                    >
                                      <glamorous(span)>
                                        <span
                                          className="glamor-3"
                                        >
                                          <glamorous(span)>
                                            <span
                                              className="glamor-1"
                                            >
                                              Menu item
                                            </span>
                                          </glamorous(span)>
                                          <glamorous(span)>
                                            <span
                                              className="glamor-2"
                                            />
                                          </glamorous(span)>
                                        </span>
                                      </glamorous(span)>
                                    </span>
                                  </glamorous(span)>
                                </div>
                              </MenuItem>
                            </MenuItem>
                            <MenuItem>
                              <MenuItem
                                disabled={undefined}
                                onClick={undefined}
                                onKeyDown={undefined}
                                tabIndex={0}
                                variant="regular"
                              >
                                <div
                                  className="glamor-5"
                                  onClick={undefined}
                                  onKeyDown={undefined}
                                  tabIndex={0}
                                >
                                  <glamorous(span)>
                                    <span
                                      className="glamor-4"
                                    >
                                      <glamorous(span)>
                                        <span
                                          className="glamor-3"
                                        >
                                          <glamorous(span)>
                                            <span
                                              className="glamor-1"
                                            >
                                              Menu item
                                            </span>
                                          </glamorous(span)>
                                          <glamorous(span)>
                                            <span
                                              className="glamor-2"
                                            />
                                          </glamorous(span)>
                                        </span>
                                      </glamorous(span)>
                                    </span>
                                  </glamorous(span)>
                                </div>
                              </MenuItem>
                            </MenuItem>
                          </div>
                        </MenuGroup>
                      </MenuGroup>
                      <MenuGroup
                        title="Menu Group Title"
                      >
                        <MenuGroup>
                          <div
                            className="glamor-11"
                          >
                            <MenuGroupTitle>
                              <MenuGroupTitle>
                                <h3
                                  className="glamor-0"
                                >
                                  Menu Group Title
                                </h3>
                              </MenuGroupTitle>
                            </MenuGroupTitle>
                            <MenuItem>
                              <MenuItem
                                disabled={undefined}
                                onClick={undefined}
                                onKeyDown={undefined}
                                tabIndex={0}
                                variant="regular"
                              >
                                <div
                                  className="glamor-5"
                                  onClick={undefined}
                                  onKeyDown={undefined}
                                  tabIndex={0}
                                >
                                  <glamorous(span)>
                                    <span
                                      className="glamor-4"
                                    >
                                      <glamorous(span)>
                                        <span
                                          className="glamor-3"
                                        >
                                          <glamorous(span)>
                                            <span
                                              className="glamor-1"
                                            >
                                              Menu item
                                            </span>
                                          </glamorous(span)>
                                          <glamorous(span)>
                                            <span
                                              className="glamor-2"
                                            />
                                          </glamorous(span)>
                                        </span>
                                      </glamorous(span)>
                                    </span>
                                  </glamorous(span)>
                                </div>
                              </MenuItem>
                            </MenuItem>
                            <MenuItem>
                              <MenuItem
                                disabled={undefined}
                                onClick={undefined}
                                onKeyDown={undefined}
                                tabIndex={0}
                                variant="regular"
                              >
                                <div
                                  className="glamor-5"
                                  onClick={undefined}
                                  onKeyDown={undefined}
                                  tabIndex={0}
                                >
                                  <glamorous(span)>
                                    <span
                                      className="glamor-4"
                                    >
                                      <glamorous(span)>
                                        <span
                                          className="glamor-3"
                                        >
                                          <glamorous(span)>
                                            <span
                                              className="glamor-1"
                                            >
                                              Menu item
                                            </span>
                                          </glamorous(span)>
                                          <glamorous(span)>
                                            <span
                                              className="glamor-2"
                                            />
                                          </glamorous(span)>
                                        </span>
                                      </glamorous(span)>
                                    </span>
                                  </glamorous(span)>
                                </div>
                              </MenuItem>
                            </MenuItem>
                          </div>
                        </MenuGroup>
                      </MenuGroup>
                    </div>
                  </Menu>
                </Menu>
              </div>
            </glamorous(div)>
          </div>
        </LivePreview>
      </div>
    </LiveProvider>
  </ThemeProvider>
</ThemeProvider>
`;
